From a3115e42eadaf9cec590fb5d692c08f75e841b9c Mon Sep 17 00:00:00 2001
From: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
Date: Thu, 6 Jul 2017 10:00:05 +0800
Subject: [PATCH] check update status for OTA rollback

check status of update, then change the status. if update has failed,
use ramdisk filesystem to rollback it.

Signed-off-by: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
---
 common/main.c | 48 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 48 insertions(+)

diff --git a/common/main.c b/common/main.c
index 2116a9e..3343bfb 100644
--- a/common/main.c
+++ b/common/main.c
@@ -12,6 +12,8 @@
 #include <cli.h>
 #include <console.h>
 #include <version.h>
+#include <mmc.h>
+#include <asm/io.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
@@ -40,6 +42,50 @@ static void run_preboot_environment_command(void)
 #endif /* CONFIG_PREBOOT */
 }
 
+struct sdversion_t{
+	unsigned char   updateflag;
+	unsigned char   updatepart;
+	unsigned char   data[0x200 - 2];
+};
+
+int check_SDenv_version(void)
+{
+	struct mmc *mmc;
+	uint blk_start, blk_cnt;
+	unsigned long offset, size;
+	struct sdversion_t sdversion_env;
+	int dev = 0;
+
+	offset	= 0x1FE000;
+	size	= 0x200;
+	mmc = find_mmc_device(dev);
+	struct blk_desc *desc = mmc_get_blk_desc(mmc);
+	blk_start   = ALIGN(offset, mmc->read_bl_len) / mmc->read_bl_len;
+	blk_cnt     = ALIGN(size, mmc->read_bl_len) / mmc->read_bl_len;
+	blk_dread(desc, blk_start, blk_cnt, (uchar *)&sdversion_env);
+	if(sdversion_env.updateflag == '1')
+	{
+		printf("system is updating\n");
+		sdversion_env.updateflag = '2';
+		blk_dwrite(desc, blk_start, blk_cnt, (uchar *)&sdversion_env);
+		return 0;
+	}
+	else if(sdversion_env.updateflag == '2')
+	{
+		sdversion_env.updateflag = '3';
+		blk_dwrite(desc, blk_start, blk_cnt, (uchar *)&sdversion_env);
+		if(sdversion_env.updatepart == '3')
+		{
+			setenv("bootfile","uImage_old");
+			return 0;
+		}
+		else
+			return 1;
+	}
+	else
+		return 0;
+}
+
 /* We come here after U-Boot is initialised and ready to process commands */
 void main_loop(void)
 {
@@ -51,6 +97,8 @@ void main_loop(void)
 	setenv("ver", version_string);  /* set version variable */
 #endif /* CONFIG_VERSION_VARIABLE */
 
+	if(check_SDenv_version())
+		setenv("bootcmd","run rollbackboot");
 	cli_init();
 
 	run_preboot_environment_command();
-- 
2.1.0.27.g96db324

